packages/tree-widget/src/components/search-bar/Popup.tsx

```diff
@@ -0,0 +1,317 @@
+        const topMargin = containerStyle.marginTop
+          ? parseFloat(containerStyle.marginTop)
+          : 0;
+        if (
+          targetRect.top - popupRect.height - topMargin - offset <
+          viewportRect.top
+        ) {
+          return RelativePosition.BottomLeft;
+        }
+        break;
+      }
+
+      case RelativePosition.BottomLeft: {
+        const bottomMargin = containerStyle.marginBottom
+          ? parseFloat(containerStyle.marginBottom)
+          : 0;
+        if (
+          targetRect.bottom + popupRect.height + bottomMargin + offset >
+          viewportRect.bottom
+        ) {
+          return RelativePosition.TopLeft;
+        }
+        break;
+      }
+
+      case RelativePosition.Bottom: {
+        const bottomMargin = containerStyle.marginBottom
+          ? parseFloat(containerStyle.marginBottom)
+          : 0;
+        if (
+          targetRect.bottom + popupRect.height + bottomMargin + offset >
+          viewportRect.bottom
+        ) {
+          return RelativePosition.Top;
+        }
+        break;
+      }
+
+      case RelativePosition.Top: {
+        const topMargin = containerStyle.marginTop
+          ? parseFloat(containerStyle.marginTop)
+          : 0;
+        if (
+          targetRect.top - popupRect.height - topMargin - offset <
+          viewportRect.top
+        ) {
+          return RelativePosition.Bottom;
+        }
+        break;
+      }
+
+      case RelativePosition.Left: {
+        const leftMargin = containerStyle.marginLeft
+          ? parseFloat(containerStyle.marginLeft)
+          : 0;
+        if (
+          targetRect.left - popupRect.width - leftMargin - offset <
+          viewportRect.left
+        ) {
+          return RelativePosition.Right;
+        }
+        break;
+      }
+
+      case RelativePosition.Right: {
+        const rightMargin = containerStyle.marginRight
+          ? parseFloat(containerStyle.marginRight)
+          : 0;
+        if (
+          targetRect.right + popupRect.width + rightMargin + offset >
+          viewportRect.right
+        ) {
+          return RelativePosition.Left;
+        }
+        break;
+      }
+    }
+
+    return this.props.position;
+  }
+
+  public render(): JSX.Element {
+    const className = classnames(
+      "tree-widget-search-popup",
+      this.getPositionClassName(this.state.position),
+      this.props.showShadow && "tree-widget-search-popup-shadow",
+      this.state.isShown && "visible",
+      this.props.showArrow && "arrow",
+      this.props.className
+    );
+
+    let style: React.CSSProperties | undefined;
+    if (this.props.fixedPosition) {
+      style = {
+        top: this.props.fixedPosition.top,
+        left: this.props.fixedPosition.left,
+        position: "fixed",
+      };
+    }
+
+    return (
+      <div style={style} className={className} ref={this._popupRef}>
+        {this.props.children}
+      </div>
+    );
+  }
+}
+
+export default Popup;
```

Comment:
I guess we could say the same thing about using class vs functional components as well

Reply:
Good catch. Changed to named export for consistency as it doesn't look like there's a single default throughout the project. Looks like the only class components that exist are `IconButton` and everything under the `search-bar` folder. Converting `IconButton` is pretty simple, but since @JoeZman is moving the search bar out to somewhere else I'll wait until that's completed and then consume it instead of converting those class components to functional ones.